types/middlewares.d.ts

```diff
@@ -31,8 +31,16 @@ declare module 'mongoose' {
   type SchemaPreOptions = MiddlewareOptions;
   type SchemaPostOptions = MiddlewareOptions;
 
-  type PreMiddlewareFunction<ThisType = any> = (this: ThisType, next: CallbackWithoutResultAndOptionalError) => void | Promise<void>;
-  type PreSaveMiddlewareFunction<ThisType = any> = (this: ThisType, next: CallbackWithoutResultAndOptionalError, opts: SaveOptions) => void | Promise<void>;
+  type PreMiddlewareFunction<ThisType = any> = (
+    this: ThisType,
+    next: CallbackWithoutResultAndOptionalError,
+    opts?: Record<string, any>
+  ) => void | Promise<void>;
+  type PreSaveMiddlewareFunction<ThisType = any> = (
+    this: ThisType,
+    next: CallbackWithoutResultAndOptionalError,
+    opts: SaveOptions
```

Comment:
Any particular reason `opts` is not optional here, and is optional in `PreMiddlewareFunction`?
Does it make any difference when defining a last parameter for a function?

Reply:
For `pre('save')`, Mongoose defaults to passing empty options, so for `schema.pre('save', function(next, opts) { opts; })` opts will always be defined. Even if you call `save()` with no args.

Not true for all middleware functions.